### PR TITLE
[mlir][ArmSME] Add initial SME vector legalization pass

### PR DESCRIPTION
**Pull request by MacDue**
This adds a new pass (`-arm-sme-vector-legalization`) which legalizes vector operations so that they can be lowered to ArmSME. This initial patch adds decomposition for `vector.outerproduct`, `vector.transfer_read`, and `vector.transfer_write` when they operate on vector types larger than a single SME tile. For example, a [8]x[8]xf32 outer product would be decomposed into four [4]x[4]xf32 outer products, which could then be lowered to ArmSME. These three ops have been picked as supporting them alone allows lowering matmuls that use all ZA accumulators to ArmSME.

For it to be possible to legalize a vector type it has to be a multiple of an SME tile size, but other than that any shape can be used. E.g. `vector<[8]x[8]xf32>`, `vector<[4]x[16]xf32>`, `vector<[16]x[4]xf32>` can all be lowered to four `vector<[4]x[4]xf32>` operations.

In future, this pass will be extended with more SME-specific rewrites to legalize unrolling the reduction dimension of matmuls (which is not type-decomposition), which is why the pass has quite a general name.

### CODE REVIEW

**Comment by MacDue** [comment]
Note: The tests passing depend on #78983.

File: mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+namespace {
+
+struct SMETile {
+  // Note: The units of (row, col) are vscale (as SME tiles are scalable).
+  int row{0};
+  int col{0};
+  VectorType type;
+};
+
+/// Adds a constant scalable offset to `indices`. i.e. for 2D:
```
Comment (banach-space) [inline mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp:29]:
```suggestion
/// Adds. elementwise, `scalableOffset` to `indices`. i.e. for 2D:
```
? Also, are `indices` required to be same length as `scalableOffset`?

File: mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+/// { indices[0] + offset[0] * vscale, indices[1] + offset[1] * vscale }
+SmallVector<Value, 2> addConstantScalableOffset(OpBuilder &builder,
+                                                Location loc,
+                                                ValueRange indices,
+                                                ArrayRef<int> scalableOffset) {
```
Comment (banach-space) [inline mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp:34]:
```suggestion
                                                ArrayRef<int> scalableOffsets) {
```
?

File: mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+  return llvm::map_to_vector(
+      llvm::zip_equal(indices, scalableOffset), [&](auto pair) -> Value {
+        auto [index, base] = pair;
+        auto offset = builder.create<arith::MulIOp>(
+            loc, builder.create<arith::ConstantIndexOp>(loc, base), vscale);
+        return builder.create<arith::AddIOp>(loc, index, offset);
+      });
+}
+
+/// Remaps indices (e.g. from a load/store) for a larger vector type to indices
```
Comment (banach-space) [inline mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp:45]:
```suggestion
/// Remaps `indices` (e.g. from a load/store) for a larger vector type to indices
```

File: mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+/// for one of the SME tiles it will decompose into.
+SmallVector<Value, 2> remapIndicesForSMETile(OpBuilder &builder, Location loc,
+                                             ValueRange indices,
+                                             SMETile tileTile) {
+  return addConstantScalableOffset(builder, loc, indices,
+                                   {tileTile.row, tileTile.col});
+}
+
+/// Returns true if `mask` is generated by an operation that can be decomposed
+/// for SME. Currently, that is just no mask, or vector.create_mask.
```
Comment (banach-space) [inline mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp:55]: Why not `vector.constant_mask`?

File: mlir/include/mlir/Dialect/ArmSME/Transforms/Passes.td [hunk 5]
```diff
@@ -122,4 +122,23 @@ def TileAllocation
   let dependentDialects = ["func::FuncDialect"];
 }
 
+def VectorLegalization
+  : Pass<"arm-sme-vector-legalization", "mlir::ModuleOp"> {
+  let summary = "Legalize vectors for ArmSME";
+  let description = [{
+    This pass legalizes vector operations so that they can be lowered to ArmSME.
+    This includes decomposing operations that operate on vector types larger
+    than a single SME tile (e.g. `vector<[8]x[8]xf32>`) into multiple SME
```
Comment (banach-space) [inline mlir/include/mlir/Dialect/ArmSME/Transforms/Passes.td:131]: `vector<[8]x[8]xf32>` matches the size of `ZA`, right? How about vectors that are larger than this? Or smaller?

File: mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+  // from the mask operands to get the parameters for this tile tile.
+  auto tileMaskDims = addConstantScalableOffset(
+      builder, loc, createMask.getOperands(), {-tileTile.row, -tileTile.col});
+  auto createTileMask = builder.create<vector::CreateMaskOp>(
+      loc, tileTile.type.clone(builder.getI1Type()), tileMaskDims);
+  return createTileMask.getResult();
+}
+
+/// Constructs an iterator that returns each SME tile (with coordinates)
+/// contained within a VectorType.
```
Comment (banach-space) [inline mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp:78]: Could you add an example?

File: mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+      [=](auto indices) {
+        int row = int(indices[0]);
+        int col = int(indices[1]);
+        if (transposeIndices)
+          std::swap(row, col);
+        return SMETile{row, col, smeTileType};
+      });
+}
+
+/// Returns the number of SME tiles that fit into the a vector type.
```
Comment (banach-space) [inline mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp:95]:
```suggestion
/// Returns the number of SME tiles that fit into `type`.
```

File: mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+int getNumberOfSMETilesForVectorType(VectorType type) {
+  assert(isMultipleOfSMETileVectorType(type));
+  int64_t vectorRows = type.getDimSize(0);
+  int64_t vectorCols = type.getDimSize(1);
+  auto elementType = type.getElementType();
+  unsigned minNumElts = getSMETileSliceMinNumElts(elementType);
+  return (vectorRows * vectorCols) / (minNumElts * minNumElts);
```
Comment (banach-space) [inline mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp:102]: What happens if this is not an integer?

File: mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp [hunk 9]
```diff
@@ -0,0 +1,308 @@
...
+        auto [index, base] = pair;
+        auto offset = builder.create<arith::MulIOp>(
+            loc, builder.create<arith::ConstantIndexOp>(loc, base), vscale);
+        return builder.create<arith::AddIOp>(loc, index, offset);
+      });
+}
+
+/// Remaps indices (e.g. from a load/store) for a larger vector type to indices
+/// for one of the SME tiles it will decompose into.
+SmallVector<Value, 2> remapIndicesForSMETile(OpBuilder &builder, Location loc,
```
Comment (banach-space) [inline mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp:47]: One thing that confuses me ...

Re-mapping indices for a larger vector to indices for a smaller vector should mean that the indices are reduced: a large vector will have a range of e.g. [0:8xVL], smaller vector will have a range of [0:4xVL]. However, it looks like `addConstantScalableOffset` increases the indices rather than "clamping" them.

I think that we are missing some examples :)

**Review by c-rhodes — COMMENTED** [review]
thanks for patch Ben this is good work! Not got through all of it yet, but left a few comments for now. Will take another pass later

File: mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp (hunk 8 above)
Comment (MacDue) [reply mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp:102]: It always will be an integer (see the `assert(isMultipleOfSMETileVectorType(type))` on entry to this function)

File: mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp (hunk 9 above)
Comment (MacDue) [reply mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp:47]: Decomposing a larger type into smaller types means the indices get bigger, see this diagram: 
```
// Decompose a 8x8 load at [a,b] into four 4x4 loads:
// initial indices = [a,b], size = 8x8
// ┌─────────────┬─────────────┐
// │[a,b]        │[a,b+4]      │
// │             │             │
// │             │             │
// │             │             │
// │             │             │
// │             │             │
// │             │             │
// ├─────────────┼─────────────┤
// │[a+4,b]      │[a+4,b+4]    │
// │             │             │
// │             │             │
// │             │             │
// │             │             │
// │             │             │
// │             │             │
// └─────────────┴─────────────┘
```

File: mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp (hunk 2 above)
Comment (MacDue) [reply mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp:34]: It's semantics, but to me you have your `indices` and you're adding a multi-dimensional `scalableOffset` offset to them (not multiple independent `scalableOffsets`)

File: mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp (hunk 1 above)
Comment (MacDue) [reply mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp:29]: They're the same length (`zip_equal` asserts that)

File: mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp [hunk 10]
```diff
@@ -0,0 +1,308 @@
...
+  assert(isMultipleOfSMETileVectorType(type));
+  int64_t vectorRows = type.getDimSize(0);
+  int64_t vectorCols = type.getDimSize(1);
+  auto elementType = type.getElementType();
+  unsigned minNumElts = getSMETileSliceMinNumElts(elementType);
+  return (vectorRows * vectorCols) / (minNumElts * minNumElts);
+}
+
+/// Legalize `vector.outerproduct` operations to fit within SME tiles.
+struct LegalizeVectorOuterProductOp
```
[the remark this answers is not in the document]
Comment (MacDue) [reply mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp:106]: Expansion/promotion to me sounds like taking a smaller type and increasing it to a larger size. I think the term for these rewrites would be decomposition.

[the remark this answers is not in the document]
Comment (MacDue) [reply mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp]: I've updated the names/docs to include decomposition.

File: mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp (hunk 2 above)
Comment (banach-space) [reply mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp:51]: Is this an elementwise operation? As in, for every "index" in `indices` there's an "offset" in `scalableOffset` and the result is "index" + "offset"? If yes then this will basically be adding 2 vectors and:
* `indices` implies that there are multiple elements in this vector,
* `scalableOffset` implies that there's only one element.

File: mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp [hunk 11]
```diff
@@ -0,0 +1,308 @@
...
+#define GEN_PASS_DEF_VECTORLEGALIZATION
+#include "mlir/Dialect/ArmSME/Transforms/Passes.h.inc"
+} // namespace mlir::arm_sme
+
+using namespace mlir;
+using namespace mlir::arm_sme;
+
+namespace {
+
+struct SMETile {
```
Comment (banach-space) [inline mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp:38]: Please include some ascii that would explain the meaning of `row` and `col` in this context. Also, IMHO this isn't really `SMETile` - it's more like `SubTileCoordinates`.

File: mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp [hunk 12]
```diff
@@ -0,0 +1,308 @@
...
+        int col = int(indices[1]);
+        if (transposeIndices)
+          std::swap(row, col);
+        return SMETile{row, col, smeTileType};
+      });
+}
+
+/// Returns the number of SME tiles that fit into the a vector type.
+int getNumberOfSMETilesForVectorType(VectorType type) {
+  assert(isMultipleOfSMETileVectorType(type));
```
Comment (banach-space) [inline mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp:129]: Please add a nice error msg :)

File: mlir/include/mlir/Dialect/ArmSME/Transforms/Passes.td (hunk 5 above)
Comment (banach-space) [reply mlir/include/mlir/Dialect/ArmSME/Transforms/Passes.td:131]: Right - but are there any limitations? I know that there are -  those are effectively enforced by `isMultipleOfSMETileVectorType`, right? It would be good add a few more comments here, e.g.:
* "legal" in this context means something that matches SMEs virtual tiles as available in hardware
* "input" type must be a multiple of these tiles (so definitely 2d and scalable in both dimensions)

File: mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp [hunk 13]
```diff
@@ -0,0 +1,308 @@
...
+            loc, builder.create<arith::ConstantIndexOp>(loc, base), vscale);
+        return builder.create<arith::AddIOp>(loc, index, offset);
+      });
+}
+
+/// Remaps indices (e.g. from a load/store) for a larger vector type to indices
+/// for one of the SME tiles it will decompose into.
+SmallVector<Value, 2> remapIndicesForSMETile(OpBuilder &builder, Location loc,
+                                             ValueRange indices,
+                                             SMETile tileTile) {
```
Comment (banach-space) [inline mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp:78]: We need a more descriptive name for this:
* `subTileId`
* `subTileSpec`
* `subTileCoordinates`
* `subTileOffsets`

**Review by banach-space — COMMENTED** [review]
Posting these comments now as I need to be AFK for a while.

My general impression is that this is missing a clearly defined input and output formats. In the context of this pass, what is a tile?
* SME tile that can me mapped onto hardware virtual tile?
* Linalg tile (as in a generic tile after Linalg tiling)?

What do we get after this decomposition? Also tiles? Perhaps something along the lines:
> Maps "generic" 2D scalable tiles (as produced by the Linalg tiling and vectorisation logic) into "SME" 2D tiles (smaller 2D scalable tiles) that match the size of SME virtual tiles (as available in hardware).

Basically, variable names like `tileTile` are a hint that we need to expand the dictionary a bit :-) Perhaps just use `GenericTile` and `SMESubTile`?

**Comment by MacDue** [comment]
> Basically, variable names like tileTile are a hint that we need to expand the dictionary a bit :-) Perhaps just use GenericTile and > SMESubTile?

`tileTile` is just rename typo that meant to be called `subTile` or ` smeTile`

**Comment by MacDue** [comment]
> Perhaps just use GenericTile and SMESubTile?
(all uses of the word 'tile' were intended to refer to SME tiles, and the generic case is just a `vectorType`).

**Comment by banach-space** [comment]
> > Perhaps just use GenericTile and SMESubTile?
> 
> (all uses of the word 'tile' were intended to refer to SME tiles, and the generic case is just a `vectorType`).

Got it - I still think it would be good to use stricter names. For somebody focusing on SME, "tile" will normally mean "SME virtual tile". So all good. However, for people more used to Linalg level of abstractions, "tile" is just a slice of the iteration space. I just think that the term is a bit ambiguous and overloaded.

Another option that could help disambiguate:
* `ZATile` to represent e.g. `vector<[8]x[8]xi32>`, and 
* `ZASubTile` for things like `vector<[4]x[4]xi32>`.

These would be different for different element types.

File: mlir/include/mlir/Dialect/ArmSME/Transforms/Passes.td (hunk 5 above)
Comment (banach-space) [reply mlir/include/mlir/Dialect/ArmSME/Transforms/Passes.td:131]: > legal - means something that can be lowered to ArmSME 

Perhaps you could try to defining that? I assume it means either:
* 2d scalable vectors with sizes matching SME virtual tiles,
* 1d scalable vectors
* n-D vectors with at most the trailing dim scalable?

**Comment by MacDue** [comment]
I've made added some more context to the header of the pass, that within this context `tile` always means SME tile (and renamed most things to say `smeTile`).  

> Another option that could help disambiguate:
> 
> * `ZATile` to represent e.g. `vector<[8]x[8]xi32>`, and
> * `ZASubTile` for things like `vector<[4]x[4]xi32>`.
> 
> These would be different for different element types.

This pass is not considering ZA, the input `vectorType` does not have map to the size of ZA, it can be bigger, smaller, or the same size, but that's not the concern of this pass. This pass just wants to convert operations on `vectorTypes` that are some multiple of SME tiles, into multiple SME-sized operations.

File: mlir/test/Integration/Dialect/Linalg/CPU/ArmSME/multi-tile-matmul.mlir [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+}
+
+module attributes {transform.with_named_sequence} {
+  transform.named_sequence @__transform_main(%module : !transform.any_op {transform.consumed}) {
+    %matmul = transform.structured.match ops{["linalg.matmul"]} in %module
+      : (!transform.any_op) -> !transform.any_op
+
+    // Step 1: Tile for size [8] x [8], which corresponds to (2 x SVLs) x (2 x SVLs),
+    // where SVLs is the number of 32-bit elements in a vector of SVL bits.
+    // This uses all four 32-bit SME virtual tiles.
```
Comment (banach-space) [inline mlir/test/Integration/Dialect/Linalg/CPU/ArmSME/multi-tile-matmul.mlir:72]: I think that this is key and I would emphasis this a bit better, e.g.:
> NOTE: We are using tile sizes larger than SME virtual tiles. This is specifically to exercise the ArmSME type legalization pass.

Or something similar :) Otherwise it's not clear what's being tested. Perhaps adding such note at the top would be better 🤔 .

File: mlir/test/Integration/Dialect/Vector/CPU/ArmSME/test-multi-tile-transpose.mlir [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+  %c128 = arith.constant 128 : i32
+  func.call @setArmSVLBits(%c128) : (i32) -> ()
+
+  //      CHECK: Input memref:
+  //      CHECK:  [1,   1,   1,   1,   1,   1,   1,   1,   1,   1,   1,   1,   1,   1,   1,   1]
+  // CHECK-NEXT:  [2,   2,   2,   2,   2,   2,   2,   2,   2,   2,   2,   2,   2,   2,   2,   2]
+  // CHECK-NEXT:  [3,   3,   3,   3,   3,   3,   3,   3,   3,   3,   3,   3,   3,   3,   3,   3]
+  // CHECK-NEXT:  [4,   4,   4,   4,   4,   4,   4,   4,   4,   4,   4,   4,   4,   4,   4,   4]
+  //
+  //      CHECK: (Masked 15x2) transposed result:
```
Comment (banach-space) [inline mlir/test/Integration/Dialect/Vector/CPU/ArmSME/test-multi-tile-transpose.mlir:123]: [nit] IMHO, this would be a bit easier to reason about:
```
  //      CHECK: (Masked 15x2) transposed result:
  // ....
  %maskRows = arith.constant 2 : index
  %maskCols = arith.constant 15 : index
  func.call @testTransposedReadWithMask(%maskRows, %maskCols) : () -> ()
```

Similar comment for other tests.

File: mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp [hunk 16]
```diff
@@ -0,0 +1,379 @@
...
+/// [4]x[4] sub-tiles.
+///
+///           8 x vscale
+/// ┌─────────────┬─────────────┐
+/// │(0,0)        │(0,4)        │
+/// │             │             │
+/// ├─────────────┼─────────────┤ 8 x vscale
+/// │(4,0)        │(4,4)        │
+/// │             │             │
+/// └─────────────┴─────────────┘
```
Comment (banach-space) [inline mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp:58]: Love this! 

> For example for an [8]x[8] tile would have four [4]x[4] sub-tiles.

Should this mention that it's an example specifically for 32-bit elements? I guess the intent is to keep this fairly abstract, so [8]x[8] could, in theory, be decomposed into 16 [2]x[2] tiles? I think that you can just replace:

> For example for an [8]x[8] tile would have four [4]x[4] sub-tiles.

with e.g.

> For example for an [8]x[8] tile with four [4]x[4] sub-tiles, we would have:

and then it should work for all cases.

File: mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp (hunk 4 above)
Comment (banach-space) [reply mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp:106]: Please add a TODO (e.g. "Add support for vector.constant_mask once required for SME"). Otherwise it's not clear whether this limitation is by design, due to some other limitations etc ...

File: mlir/test/Dialect/ArmSME/vector-legalization.mlir [hunk 17]
```diff
@@ -0,0 +1,268 @@
...
+  // CHECK-DAG: %[[RES_3:.*]] = vector.mask %[[MASK_3]] { vector.outerproduct %[[LHS_3]], %[[RHS_0]] : vector<[4]xf32>, vector<[4]xf32> } : vector<[4]x[4]xi1> -> vector<[4]x[4]xf32>
+  // CHECK-NEXT: return %[[RES_0]], %[[RES_1]], %[[RES_2]], %[[RES_3]] : vector<[4]x[4]xf32>, vector<[4]x[4]xf32>, vector<[4]x[4]xf32>, vector<[4]x[4]xf32>
+  %mask = vector.create_mask %lhs_dim, %rhs_dim : vector<[16]x[4]xi1>
+  %0 = vector.mask %mask { vector.outerproduct %lhs, %rhs : vector<[16]xf32>, vector<[4]xf32> } : vector<[16]x[4]xi1> -> vector<[16]x[4]xf32>
+  return %0 : vector<[16]x[4]xf32>
+}
+
+// -----
+
+/// This demonstrates a rectangular tiling that uses all f64 accumulators.
```
Comment (banach-space) [inline mlir/test/Dialect/ArmSME/vector-legalization.mlir:89]: [nit] For me "rectangular" would be more like:
```mlir
func.func @outerproduct_f64_scalable_8x4_no_acc(%lhs: vector<[8]xf64>, %rhs: vector<[2]xf64>) -> vector<[8]x[2]xf64>
```

As in, as "rectangular" as possible :) Perhaps worth testing a corner case instead?

File: mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp [hunk 18]
```diff
@@ -0,0 +1,379 @@
...
+          loc, smeTileType, lhs, rhs,
+          !accSMETiles.empty() ? accSMETiles[index] : Value{},
+          outerProductOp.getKind());
+
+      auto maskedOuterProduct =
+          vector::maskOperation(rewriter, smeOuterProduct, smeMask);
+      resultSMETiles.push_back(maskedOuterProduct->getResult(0));
+    }
+
+    rewriter.replaceOp(rootOp, resultSMETiles, adaptor.getResultMapping());
```
Comment (banach-space) [inline mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp:211]: Could you test/demonstrate how users of `rootOp` are updated? Across function boundaries :)

**Review by banach-space — COMMENTED** [review]
The extra comments are super helpful, thank you Ben!

File: mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp (hunk 18 above)
Comment (MacDue) [reply mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp:211]: Not much more than is already shown within the `vector-legalization.mlir` test (where users such as returns, stores, etc) are updated. The 1-to-N conversion would fail if any intermediate unrealized conversions were left. 

The magic of gluing these rewrites together is all handled automatically by the 1-to-N conversion framework.

File: mlir/test/Dialect/ArmSME/vector-legalization.mlir (hunk 17 above)
Comment (MacDue) [reply mlir/test/Dialect/ArmSME/vector-legalization.mlir:89]: Rectangular was called out in that test as it's demonstrating a full-ZA f64 tiling, which has no square tiling, which makes it slightly interesting when it comes to tiling for it.

Comment (MacDue) [reply mlir/test/Dialect/ArmSME/vector-legalization.mlir:89]: There are no other tests this shape in terms of tiles, i.e. multiple tiles for both the rows/cols, but not an equal number of tiles. There are already a few tests that have shapes that are (n tiles) x (1 tiles), which is what [8]x[2] is (it's 4x1 f64 tiles), whereas [8]x[4] is 4x2 tiles.

File: mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp (hunk 9 above)
Comment (MacDue) [reply mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp]: I hope you don't think I intentionally named somthing `tileTile` :laughing:, I spent a while renaming things, and messed up a global find and replace :sweat_smile:

File: mlir/test/Integration/Dialect/Linalg/CPU/ArmSME/multi-tile-matmul.mlir (hunk 14 above)
Comment (MacDue) [reply mlir/test/Integration/Dialect/Linalg/CPU/ArmSME/multi-tile-matmul.mlir:78]: (added a summary of the test at the top of the file)

**Review by banach-space — APPROVED** [review]
Thanks for addressing my comments - LGTM!

**Comment by dcaballe** [comment]
Great work! I acknowledge the complexity of this PR but I was wondering how much the "vector register splitting" side could be generalized to make it reusable by other targets. Vector register legalization is something that has come up multiple times so having a pass that can be parametrized accordingly would be extremely valuable. What are your thoughts about this? :)

**Review by c-rhodes — APPROVED** [review]
great work Ben this is very well written! LGTM cheers

**Comment by c-rhodes** [comment]
> Great work! I acknowledge the complexity of this PR but I was wondering how much the "vector register splitting" side could be generalized to make it reusable by other targets. Vector register legalization is something that has come up multiple times so having a pass that can be parametrized accordingly would be extremely valuable. What are your thoughts about this? :)

I agree, made a similar observation offline and think we should look into this as a follow up.

**Comment by MacDue** [comment]
> Great work! I acknowledge the complexity of this PR but I was wondering how much the "vector register splitting" side could be generalized to make it reusable by other targets. Vector register legalization is something that has come up multiple times so having a pass that can be parametrized accordingly would be extremely valuable. What are your thoughts about this? :)

I think the logic here could be generalized (though, as written, I've made assumptions about scalability). I think it'd always be a rank-preserving legalization (i.e. 2D -> 2D, like here, or 1D to 1D, etc). Other transforms like unrolling fixed vectors can already be handled elsewhere (and would be complex to incorporate).

File: mlir/lib/Dialect/ArmSME/Transforms/TileAllocation.cpp
[the remark this answers is not in the document]
Comment (MacDue) [reply mlir/lib/Dialect/ArmSME/Transforms/TileAllocation.cpp:1]: Not sure how that happened (sometimes my editor does random stuff :sweat_smile:)
